ompi/mca/coll/ucc/coll_ucc_component.c

```diff
@@ -0,0 +1,173 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2021 Mellanox Technologies. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+#include "ompi_config.h"
+#include "coll_ucc.h"
+#include "coll_ucc_dtypes.h"
+#include "opal/util/argv.h"
+
+static int mca_coll_ucc_open(void);
+static int mca_coll_ucc_close(void);
+static int mca_coll_ucc_register(void);
+
+int mca_coll_ucc_output = -1;
+
+mca_coll_ucc_component_t mca_coll_ucc_component = {
+    /* First, the mca_component_t struct containing meta information
+       about the component  */
+    {
+        .collm_version = {
+            MCA_COLL_BASE_VERSION_2_4_0,
+
+            /* Component name and version */
+            .mca_component_name = "ucc",
+            MCA_BASE_MAKE_VERSION(component, OMPI_MAJOR_VERSION, OMPI_MINOR_VERSION,
+                                  OMPI_RELEASE_VERSION),
+
+            /* Component open and close functions */
+            .mca_open_component            = mca_coll_ucc_open,
+            .mca_close_component           = mca_coll_ucc_close,
+            .mca_register_component_params = mca_coll_ucc_register,
+            .mca_query_component           = NULL,
+        },
+        .collm_data = {
+            /* The component is not checkpoint ready */
+            MCA_BASE_METADATA_PARAM_NONE
+        },
+
+        /* Initialization / querying functions */
+        .collm_init_query = mca_coll_ucc_init_query,
+        .collm_comm_query = mca_coll_ucc_comm_query,
+    },
+    100,               /* ucc_priority                */
+    0,                 /* ucc_verbose                 */
+    1,                 /* ucc_enable                  */
+    2,                 /* ucc_np                      */
+    "basic",           /* cls                         */
+    COLL_UCC_CTS_STR,  /* requested coll_types string */
+    UCC_VERSION_STRING /* ucc version                 */
+};
+
+static int mca_coll_ucc_register(void)
+{
+    mca_coll_ucc_component_t *cm = &mca_coll_ucc_component;
```

Comment:
Is this struct initialized?

Reply:
Right above it

Reply:
Oh jeez, I am blind. Thanks.

Possibly another stupid question, the `ucc_priority` var is set to 100. Will that force it above all other collectives?

Reply:
Right, I guess by default the component will not be built unless libucc is present, those who build with it should have to increase priority in addition to building ompi with libucc.
@vspetrov thoughts? Maybe we should lower priority initially.

Reply:
Yes, the original logic was: if someone builds OMPI with libucc then it means he wants to use it.

Reply:
Just curious: suppose I am using a system-installed OMPI that defaults to using UCX. For whatever reason, I decide I want to use TCP, and so I force selection of the "ob1" PML and the "tcp" BTL.

How does this component know to disable itself? Note that I (as a simple user) may have no idea this component even exists - a user shouldn't be required to read the OMPI code base just to figure out how to run their app.

Reply:
@rhc54 how is this component related directly to ucx/tcp (ie, p2p selection)? In other words why "it has to disable itself in such case" ?

Reply:
Probably a poor assumption on my part - I assumed that this component was related to the UCX library, and hence used UCX-related transports. If I select "tcp" and don't want the UCX library used (i.e., I want to use the tcp BTL), then I wouldn't want to use the UCX-related collectives either. In other words, if I say "don't use UCX", I expect to not be using anything involving UCX.

Reply:
> Yes, the original logic was: if someone builds OMPI with libucc then it means he wants to use it.

This is not a good assumption. Packagers tend to build Open MPI with as many packages that they can in order to make a "full featured" Open MPI.